File: Scala/compoundlab/src/main/scala/Observable.scala
import scala.collection.mutable.ArrayBuffer

class Observable(observable : IQuackableObservable) extends IQuackableObservable {
	private val observers : ArrayBuffer[IObserver] = ArrayBuffer[IObserver]()
	private val name = "Observable"

	override def registerObserver(newObserver : IObserver) : Unit = {
		observers.append(newObserver)
	}

	override def notifyObservers : String = {
		val sb = new StringBuilder

		observers.foreach(observer => {
			sb.append(observer.update(observable) + "\n")
		})
		sb.toString.trim
	}

	override def getName: String = {
		this.name
	}
}

File: Java/compoundlab/src/main/java/com/bbd/abstractFactoryPattern/Observable.java
package com.bbd.abstractFactoryPattern;

import java.util.ArrayList;
import java.util.List;
import java.util.Observer;

public class Observable implements IQuackObservable
{
	IQuackObservable _observable;
	ArrayList<IObserver> observers = new ArrayList<>();

	Observable(IQuackObservable observable)
	{
		this._observable = observable;
	}

	public void registerObserver(IObserver observer)
	{
		observers.add(observer);
	}

	public String notifyObservers()
	{
		StringBuilder sb = new StringBuilder();
		for (IObserver observer : observers)
		{
			sb.append(observer.update(this._observable) + "\n");
		}
		return sb.toString().trim();
	}

	public String getName() { return ""; }
}
